pkg/controller/awsfederatedaccountaccess/awsfederatedaccountaccess_controller.go

```diff
@@ -0,0 +1,368 @@
+		client: mgr.GetClient(),
+		scheme: mgr.GetScheme(),
+	}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("awsfederatedaccountaccess-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource AWSFederatedAccountAccess
+	err = c.Watch(&source.Kind{Type: &awsv1alpha1.AWSFederatedAccountAccess{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	// watch for changes to secondary resource AWSFederatedRole
+	err = c.Watch(&source.Kind{Type: &awsv1alpha1.AWSFederatedRole{}}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &awsv1alpha1.AccountPool{},
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// blank assignment to verify that ReconcileAWSFederatedAccountAccess implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileAWSFederatedAccountAccess{}
+
+// ReconcileAWSFederatedAccountAccess reconciles a AWSFederatedAccountAccess object
+type ReconcileAWSFederatedAccountAccess struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a AWSFederatedAccountAccess object and makes changes based on the state read
+// and what is in the AWSFederatedAccountAccess.Spec
+func (r *ReconcileAWSFederatedAccountAccess) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling AWSFederatedAccountAccess")
+
+	// Fetch the AWSFederatedAccountAccess instance
+	currentFAA := &awsv1alpha1.AWSFederatedAccountAccess{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, currentFAA)
+	if err != nil {
+		if k8serr.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	// If the state is applied don't do anything
+	if currentFAA.Status.State == awsv1alpha1.AWSFederatedAccountStateReady || currentFAA.Status.State == awsv1alpha1.AWSFederatedAccountStateFailed {
+		return reconcile.Result{}, nil
+	}
+
+	// Get a list of all available roles
+	federatedRoleList := &awsv1alpha1.AWSFederatedRoleList{}
+	if r.client.List(context.TODO(), &client.ListOptions{}, federatedRoleList); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Create map to make finding role easier
+	roleMap := make(map[string]awsv1alpha1.AWSFederatedRole)
```

Comment:
This could probably be used elsewhere. Maybe could be moved to util function.

Reply:
One of the cards coming out of this is to move the shared functionally to a util package there are more functions that should be moved as well.

Reply:
You should be able to use this when you call `r.client.List()` to return the single role. That will eliminate some the code below. 

`&client.ListOptions{FieldSelector: fields.Set{"metadata.name": currentFAA.Spec.AWSFederatedRoleName}.AsSelector().String()}`